syntax = "proto3";

message InventoryRecord {
    string id = 1;
    string name = 2;
    string description = 3;
    double unit_price = 4;
    int32 quantity_in_stock = 5;
    double inventory_value = 6;
    int32 reorder_level = 7;
    int32 reorder_time_in_days = 8;
    int32 quantity_in_reorder = 9;
    bool discontinued = 10;
}

message InventoryRequest {
    string id = 1;
}

message InventorySearchRequest {
    string key_name = 1;
    string key_value_start = 2;
    string key_value_end = 3;
}

message PercentileResponse {
    double value = 1;
}

message UpdateRequest {
    string key_name = 1;
    string key_value = 2;
    string val_name = 3;
    string val_val_new = 4;
}

message UpdateResponse {
    bool success = 1;
}

service InventoryService {
    rpc searchByID(InventoryRequest) returns (InventoryRecord);
    rpc search(InventorySearchRequest) returns (InventoryRecord);
    rpc getDistribution(InventoryRequest) returns (PercentileResponse);
    rpc update(UpdateRequest) returns (UpdateResponse);
}
